### ISSUE TITLE
Kerbals unaffected by BDAc products

### ISSUE THREAD

**Issue by SpannerMonkey** [PapaJoesSoup/BDArmory#109]
AS above, i believe this is a knock on effect from the fix squad applied to kerbals in order for them to be safe around wheels .
https://www.youtube.com/watch?v=LAt2YL7wzbQ

**Comment by Guard13007** [comment]
@SpannerMonkey You broke your URL somehow... Seems you have `[https://www.youtube.com/watch?v=LAt2YL7wzbQ](url)` instead of just `https://www.youtube.com/watch?v=LAt2YL7wzbQ` which would automatically turn into a clickable URL. xD [👍](https://www.youtube.com/watch?v=LAt2YL7wzbQ).

**Comment by gomker** [comment]
We should be able to detect kerbal hit somehow as they are "special" vessels and modify the damage code to interact similar to how destructible buildings are handled.

**Comment by Designer225** [comment]
The Kerbals are still unaffected by BDAc (it appears that they went through the Kerbals dealing no damage!). I enabled debug lines, and the hit does not even register on the Kerbal.

I'm going to compile the mod to a branch with a debug line near where a raycast hit is being detected.

**Comment by Designer225** [comment]
I ran some tests, and it appears that Kerbals don't have any part-based colliders anymore.

The branch that I'm going to publish has some codes added for debugging. The code attempts to find a Kerbal in EVA when a bullet is fired.

When I ran the test, I managed to find Bob's part. However, when I check for the colliders, they do not exist. (If they exist, then each collider gets printed onto the debug console. Or should be.)

If Kerbals no longer have colliders, we need to find another way to fix this issue.

**Comment by SpannerMonkey** [comment]
Hi re Kerbals and colliders, the Kerbals were completely rebuilt and had substantial changes in order to prevent poor interaction with wheel and ground colliders.   Although i can't find the post/thread I recall some talk that the Kerbals now exist on a different layer than all other parts in the flight scene.  Unity uses layers extensively for different things, a full list of all currently  known  KSP relevant tags and layers can be found [Here](https://forum.kerbalspaceprogram.com/index.php?/topic/84273-unity-layers-and-tags-updated-09-march-2017/).  
 I strongly suspect that if re adding colliders is possible then it's very likely to have some odd effects in game. 
  I also suspect that this is not the case however, as the fairly extensive testing I carried out at the time of discovery leads me to believe it is indeed a layer problem. The item that leads me to this conclusion is the fact that it's ALL BDA products, not just projectile weapons that are ineffective .
 I constructed excessively explosive devices, equally several thousand tonnes of TNT and despite standing right next to the detonation the Kerbals remain unharmed unless struck by a flying piece of debris. Which does suggest that the Kerbals are not affected by things happening on the layers that BDA interacts with.
I have a vested interest in resolving this problem as it has effectively mothballed the BDA FPS mod ( since the first five mins of 1.2)  to which I'm, a major contributor.
If there's any assistance I can provide to anyone investigating this issue do get in touch.

**Comment by Designer225** [comment]
It would be great if we could find out the layer collision matrix settings from the compiled game. Given that Kerbals interact with parts, buildings, and planet surfaces, I wonder if setting the BDA layer to either of the two could fix the problem...

Of course, there is no telling what bugs could be unleashed by doing so.

**Comment by Designer225** [comment]
After a bit of struggle, I finally got the bullets to detect Kerbals.

By struggle, I meant trying to change the layer to detect collisions, and then later messing with the bit mask. I added some code with the bit mask to 17 << 17, and this time PoolBullet reported going through Kerbals.

I have pushed the changes onto my fork.

**Comment by Designer225** [comment]
I have killed Bob with the Vulcan.
![20170826174343_1](https://user-images.githubusercontent.com/4565785/29746155-2b10e346-8a86-11e7-9d2a-30d12b9029d9.jpg)

I'm going to test this several times with less powerful weapons, to make sure this isn't the result of loops.

**Comment by Designer225** [comment]
Okay, I got pretty much all of the weapon types working in terms of collision detection and damage dealing now.

There is one small issue: When controlling the Kerbal to be exploded, sometimes it does not due to some errors. (When _not_ controlling, it does explode, but some errors are thrown.)

I have pushed the changes into my fork (https://github.com/Designer225/BDArmory) now, I will create a new pull request and a temporary release build some time later.

**Comment by SpannerMonkey** [comment]
@Designer225 Hi I'll be giving your code a try with the custom Kerbals weapons, the pack was specifically designed to be used by Kerbals and has adjusted appropriate values.  If it seems promising I'll make sure that the coding team take a good look at it.

**Comment by Designer225** [comment]
@SpannerMonkey Go ahead. I really wanted to use BD FPS, as well.

I haven't fixed the aiming reticles yet, but I'll do this when I come back.

**Comment by SpannerMonkey** [comment]
Tested, I have some concerns that the branch you built from is not one of the active dev branches.  I had some strange graphical anomalies that i've not seen for a long time.

That aside your fix does indeed allow BDA products to interact once again with Kerbals, all FPS tested worked fine, didn't have any problem with aiming reticles, but then only testing over 100 meters or so.

I did however notice some nasty interaction between Kerbals and wheels,  for example , kerbals contacts wheel, kerbal explodes, or kerbal contacts wheel, vehicle is launched into orbit.  This is a problem not currently affecting the stock game, and if it's re appearance is tied firmly to your changes than it's a bit of a stumbling block , I'd imagine that most Kerbal combat would (once again) be accompanied by wheels and tracked vehicles( given that Kerbals can't walk around on ships etc unless landed and perfectly still.)  and having to avoid contact with wheels etc on pain of death,  would annoy users pretty quickly

**Comment by Designer225** [comment]
@SpannerMonkey 
1) Yeah I didn't build them from the dev branch. I'll pull files from that branch some time later.

2) I was looking at ModuleWeapon, and I found parts of code that may be related to aiming reticles. I think without changes the projectile prediction path will predict a landing location past the Kerbal..

3) My Kerbals don't even collide with the wheels even with my code added, and that was also true in the base game. The only mods I had besides BDArmory is USI Warp Drive, Module Manager, and Physics Range Extender. After installing Kerbal Foundries, I tested Kerbal collisions again. Again, no problems.

Which version of KSP, and what mods, are you using?
